modules/network/src/main/java/org/apache/ignite/internal/network/serialization/marshal/StructuredObjectMarshaller.java

```diff
@@ -0,0 +1,179 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.network.serialization.marshal;
+
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import org.apache.ignite.internal.network.serialization.ClassDescriptor;
+import org.apache.ignite.internal.network.serialization.FieldDescriptor;
+import org.apache.ignite.internal.network.serialization.SpecialMethodInvocationException;
+
+/**
+ * (Un)marshals objects that have structure (fields). These are {@link java.io.Serializable}s
+ * (which are not {@link java.io.Externalizable}s) and arbitrary (non-serializable, non-externalizable) objects.
+ */
+class StructuredObjectMarshaller implements DefaultFieldsReaderWriter {
+    private final TypedValueWriter valueWriter;
+    private final ValueReader<Object> valueReader;
+
+    private final Instantiation instantiation;
+
+    StructuredObjectMarshaller(TypedValueWriter valueWriter, ValueReader<Object> valueReader) {
+        this.valueWriter = valueWriter;
+        this.valueReader = valueReader;
+
+        instantiation = new BestEffortInstantiation(
+                new SerializableInstantiation(),
+                new UnsafeInstantiation()
+        );
+    }
+
+    void writeStructuredObject(Object object, ClassDescriptor descriptor, DataOutputStream output, MarshallingContext context)
+            throws MarshalException, IOException {
+        for (ClassDescriptor layer : lineage(descriptor)) {
+            writeStructuredObjectLayer(object, layer, output, context);
+        }
+    }
+
+    /**
+     * Returns the lineage (all the ancestors, from the progenitor (excluding Object) down the line, including the given class).
+     *
+     * @param descriptor class from which to obtain lineage
+     * @return ancestors from the progenitor (excluding Object) down the line, plus the given class itself
+     */
+    private List<ClassDescriptor> lineage(ClassDescriptor descriptor) {
+        List<ClassDescriptor> descriptors = new ArrayList<>();
+
+        ClassDescriptor currentDesc = descriptor;
+        while (currentDesc != null) {
+            descriptors.add(currentDesc);
+            currentDesc = currentDesc.superClassDescriptor();
+        }
+
+        Collections.reverse(descriptors);
+
+        return descriptors;
+    }
+
+    private void writeStructuredObjectLayer(Object object, ClassDescriptor layer, DataOutputStream output, MarshallingContext context)
+            throws IOException, MarshalException {
+        if (layer.hasSerializationOverride()) {
+            writeWithWriteObject(object, layer, output, context);
+        } else {
+            defaultWriteFields(object, layer, output, context);
+        }
+
+        context.addUsedDescriptor(layer);
+    }
+
+
```

Comment:
Odd newline

Reply:
Oops, those pesky blank lines... strnge it was not caught by checkstyle. Fixed.